[HUDI-1740] Fix insert-overwrite API archival

ssdong opened the pull request:
## *Tips*
- *Thank you very much for contributing to Apache Hudi.*
- *Please review https://hudi.apache.org/contributing.html before opening a pull request.*

## What is the purpose of the pull request

### Summary:
This pull request is to fix the archival logic within insert overwrite API against requested & inflight commit files.

### Issue:
`0.7.0` throw exception `Caused by: java.lang.IllegalArgumentException: Positive number of partitions required` while
`0.9.0-SNAPSHOT`(latest master branch) adds `java.util.NoSuchElementException: No value present in Option` on top of it when hudi tries to archive replace commit files(`COMPLETED`, `REQUESTED` and `INFLIGHT`)

Please checkout issue https://github.com/apache/hudi/issues/2707 and ticket https://issues.apache.org/jira/browse/HUDI-1740 for further information about the above 2 exceptions. 

The inner causes are somewhat sophisticated, and I've tried my best to understand them and applied the fix for it, instead of giving it a one-line tricky fix or patch and see errors gone. Of course, the approaches I am taking are open to discussions. 

### Fixes
1. `Positive number of partitions required` error is easier to be fixed where we just have to filter out empty `partitionToReplaceFileIds` for `COMPLETED` replace commit files within `ReplaceArchivalHelper.java`.
2.  `java.util.NoSuchElementException: No value present in Option` is much more complicated and it happened due to a call to `ClusteringUtils.getRequestedReplaceMetadata()` against _both_ `REQUESTED` and `INFLIGHT` commit files for retrieving their meta body. Now, I get the idea that we are encouraged to use existing utils classes for code reuse. However, a closer inspection upon `getRequestedReplaceMetadata` shows that clustering feature retrieves the meta for `INFLIGHT` commit file through a `REQUESTED` instant. Although this is _not_ fundamentally wrong since there is no "clustering plan" for both `REQUESTED` and `INFLIGHT` replace commit files so the outcome is the same for both, as it is also being pointed out in the comment within `getRequestedReplaceMetadata`. However, since `REQUESTED` instant is empty(there is a corresponding [ticket](https://issues.apache.org/jira/browse/HUDI-1739) for it), it generates an `Option.empty()` which is being fetched by `.get()` later which triggered the `NoSuchElementException`. What's more, it _loses_ information in `INFLIGHT` commit file when fetching via `REQUESTED` instant and so we observed in the following screenshot that:
<img width="1037" alt="Screen Shot 2021-04-07 at 2 09 07" src="https://user-images.githubusercontent.com/3754011/113918516-7e937a80-981d-11eb-84b6-e2c4bec2c3b1.png">

This does not make sense to me when we pretty much _abuse_ the `REQUESTED` concept to deal with `INFLIGHT` with `REQUESTED` itself being empty. I've taken the approach to define an extra field(placeholder) for `INFLIGHT` and reuse the `HoodieCommitMetadata` for deserializing `INFLIGHT` since they share the same structure where `COMPLETED` replace commit extends `HoodieCommitMetadata` for the extra `partitionToReplaceFileIds` field. Now, here's what I gained after adopting this strategy:
<img width="1065" alt="Screen Shot 2021-04-08 at 3 01 51" src="https://user-images.githubusercontent.com/3754011/113919657-cff03980-981e-11eb-928e-65c719d15ca5.png">

### The overall outcome after the fix on latest master branch:
<img width="924" alt="Screen Shot 2021-04-08 at 3 04 29" src="https://user-images.githubusercontent.com/3754011/113919768-ee563500-981e-11eb-80e8-fe62ba868709.png">

Let me know if there is anything I am missing.  :)

_The simplest solution to the 2nd issue is to actually having `ClusteringUtils.getRequestedReplaceMetadata` return `Option.of(new HoodieRequestedReplaceMetadata())` upon retrieving the empty `REQUESTED` replace commit file for both `REQUESTED` and `INFLIGHT`. I didn't choose to fix the problem this way fearing it's basically altering the inappropriate approach by giving it a bandage and stopping the bleeding._


## Update: Attaching testing evidence for both insert_overwrite_table and upsert(with clustering) cases. 
_Reason for testing clustering: clustering generates replace commit files as well._

Build and running Hudi locally with `spark-3.0.1-bin-hadoop2.7` and the following command
```
./bin/spark-shell \
 --packages org.apache.spark:spark-avro_2.12:3.0.1 \
 --conf 'spark.serializer=org.apache.spark.serializer.KryoSerializer' \
 --conf "spark.driver.extraJavaOptions=-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=127.0.0.1:35005" \
 --conf "spark.executor.extraJavaOptions=-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=127.0.0.1:25005" \
 --jars ~/temp/repository/org/apache/hudi/hudi-spark3-bundle_2.12/0.9.0-SNAPSHOT/hudi-spark3-bundle_2.12-0.9.0-SNAPSHOT.jar
```
`~/temp/repository/org/apache/hudi/` is path to my local jar.

### upsert(with clustering)

Script:
```
import org.apache.hudi.QuickstartUtils._
import scala.collection.JavaConversions._
import org.apache.spark.sql.SaveMode._
import org.apache.hudi.DataSourceReadOptions._
import org.apache.hudi.DataSourceWriteOptions._
import org.apache.hudi.config.HoodieWriteConfig._


val tableName = "hudi_date_mor"
val basePath = "/Users/susu.dong/Dev/clustering-test"

val dataGen = new DataGenerator
val inserts = convertToStringList(dataGen.generateInserts(10))
val df = spark.read.json(spark.sparkContext.parallelize(inserts, 2))
df.write.format("hudi").
        options(getQuickstartWriteConfigs).
        option(PRECOMBINE_FIELD_OPT_KEY, "ts").
        option(RECORDKEY_FIELD_OPT_KEY, "uuid").
        option(PARTITIONPATH_FIELD_OPT_KEY, "partitionpath").
        option(TABLE_NAME, tableName).
        option("hoodie.parquet.small.file.limit", "0").
        option("hoodie.keep.max.commits", "3").
        option("hoodie.cleaner.commits.retained", "1").
        option("hoodie.keep.min.commits", "2").
        option("hoodie.clustering.inline", "true").
        option("hoodie.clustering.inline.max.commits", "4").
        option("hoodie.clustering.plan.strategy.target.file.max.bytes", "1073741824").
        option("hoodie.clustering.plan.strategy.small.file.limit", "629145600").
        option("hoodie.clustering.plan.strategy.sort.columns", "rider,driver").
        mode(Append).
        save(basePath)
```
Hit `df.write` multiple times to surpass `keep.max.commits` and observe the clustering replace commit files being archived successfully:
<img width="1156" alt="Screen Shot 2021-04-11 at 1 33 23" src="https://user-images.githubusercontent.com/3754011/114278361-6d8b7900-9a6a-11eb-88ea-875b36b7c5c9.png">

### insert_overwrite_table

Script:
```
import org.apache.hudi.QuickstartUtils._
import scala.collection.JavaConversions._
import org.apache.spark.sql.SaveMode._
import org.apache.hudi.DataSourceReadOptions._
import org.apache.hudi.DataSourceWriteOptions._
import org.apache.hudi.config.HoodieWriteConfig._
import java.util.UUID
import java.sql.Timestamp

val tableName = "hudi_date_mor"
val basePath = "/Users/susu.dong/Dev/insert-overwrite-test"
val writeConfigs = Map(
     "hoodie.cleaner.incremental.mode" -> "true",
     "hoodie.insert.shuffle.parallelism" -> "2",
     "hoodie.upsert.shuffle.parallelism" -> "2",
     "hoodie.clean.automatic" -> "false",
     "hoodie.datasource.write.operation" -> "insert_overwrite_table",
     "hoodie.table.name" -> tableName,
     "hoodie.datasource.write.table.type" -> "MERGE_ON_READ",
     "hoodie.cleaner.policy" -> "KEEP_LATEST_FILE_VERSIONS",
     "hoodie.keep.max.commits" -> "3",
     "hoodie.cleaner.commits.retained" -> "1",
     "hoodie.keep.min.commits" -> "2",
     "hoodie.compact.inline.max.delta.commits" -> "1"
     )

val dateSMap: Map[Int, String] = Map(
    0-> "2020-07-01",
    1-> "2020-08-01",
    2-> "2020-09-01",
)
val dateMap: Map[Int, Timestamp] = Map(
    0-> Timestamp.valueOf("2010-07-01 11:00:15"),
    1-> Timestamp.valueOf("2010-08-01 11:00:15"),
    2-> Timestamp.valueOf("2010-09-01 11:00:15"),
)
var seq = Seq(
    (0, "value", dateMap(0), dateSMap(0), UUID.randomUUID.toString)
)
for(i <- 501 to 1000) {
    seq :+= (i, "value", dateMap(i % 3), dateSMap(i % 3), UUID.randomUUID.toString)
}
val df = seq.toDF("id", "string_column", "timestamp_column", "date_string", "uuid")

df.write.format("hudi").options(writeConfigs).mode(Overwrite).save(basePath)
```
Hit `df.write` multiple times to surpass `keep.max.commits` and observe insert_overwrite_table replace commit files being archived successfully:
<img width="1006" alt="Screen Shot 2021-04-11 at 2 10 44" src="https://user-images.githubusercontent.com/3754011/114278598-88aab880-9a6b-11eb-999c-35efc9e9af60.png">


## Brief change log

*(for example:)*
  - *Modify AnnotationLocation checkstyle rule in checkstyle.xml*

## Verify this pull request

*(Please pick either of the following options)*

This pull request is a trivial rework / code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

  - *Fix insert overwrite API archival*

## Committer checklist

 - [x] Has a corresponding JIRA in PR title & commit
 
 - [x] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

satishkotha commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/utils/MetadataConversionUtils.java:
```diff
@@ -105,14 +114,15 @@ public static HoodieArchivedMetaEntry createMetaWrapper(HoodieInstant hoodieInst
+    Option<byte[]> content = metaClient.getActiveTimeline().getInstantDetails(requestedInstant);
+    if (!content.isPresent() || content.get().length == 0) {
+      LOG.warn("No content found in requested file for instant " + pendingReplaceInstant);
+      return Option.of(new HoodieRequestedReplaceMetadata());
```
Why not return Option.empty() and skip archival for this case?

zherenyu831 replied:
Current logic is using `import org.apache.hudi.common.model.HoodieCommitMetadata.fromBytes()` to fetch empty deltacommit (bytes = []), and creating a new instance of metadata which bytes not empty. So I was thinking it may be better to keep same behaviour in ReplaceCommitRequestedInstant.

zherenyu831 replied:
I will try with what you suggested

satishkotha commented on hudi-common/src/main/java/org/apache/hudi/common/table/view/AbstractTableFileSystemView.java:
```diff
@@ -245,7 +245,7 @@ public final void reset() {
 
       // Initialize with new Hoodie timeline.
-      init(metaClient, getTimeline());
+      init(metaClient, metaClient.reloadActiveTimeline());
```
IIUC, this is breaking some fundamental assumptions.  There are many places where we pass "trimmed" timeline for time-travel queries etc. You are replacing that with all instants from active timeline, which is not desired.

Is this change needed if we handle empty partitionToReplaceFileIds in archival?

zherenyu831 replied:
@satishkotha 
Since this part will called after archival, and the archived commits still in the timeline.
In the post process hudi will try to load the byte from them, and will cause IO error

zherenyu831 replied:
I think the root problem is why we are calling reset() when close the timeline

satishkotha replied:
@bvaradar mentioned this is by design. Balaji, could you please help resolve this?

ssdong replied:
I've put `getTimeline()` back since it should be irrelevant to the archival issue though I am still curious to learn why this is needed.

satishkotha commented:
@ssdong  thanks for bringing this up and contributing. I added some comments, please take a look. Also, looks like there are some CI failures. Please fix those as well.

susudong commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/utils/MetadataConversionUtils.java:
```diff
@@ -72,9 +76,14 @@ public static HoodieArchivedMetaEntry createMetaWrapper(HoodieInstant hoodieInst
               .fromBytes(metaClient.getActiveTimeline().getInstantDetails(hoodieInstant).get(), HoodieReplaceCommitMetadata.class);
           archivedMetaWrapper.setHoodieReplaceCommitMetadata(ReplaceArchivalHelper.convertReplaceCommitMetadata(replaceCommitMetadata));
+        } else if (hoodieInstant.isInflight()) {
+          // inflight replacecommit files have the same meta data body as HoodieCommitMetadata
```
Thanks for pointing that out! Let me test it with clustering and see what happens. 😄

ssdong replied:
Thanks for pointing that out. Will test against clustering and see what happens. If it doesn’t work, will find an alternative way. 😄

zherenyu831 replied:
@satishkotha 
Same with deltacommit.requested, it is also empty, but when using `HoodieCommitMetadata.fromBytes() `, a new empty instance will be created. better to have a test case, but it should be fine

satishkotha replied:
ok, sounds good. I think most of unit test framework already is setup. Do you mind adding that small test with inflight empty file?

ssdong replied:
Hey @satishkotha The `inflight` was _originally_ empty in the unit tests before I _change_ it to have meta body in it. I wasn’t seeing the whole picture that `inflight` could be both empty and non-empty, neither were the original unit tests covering both cases. We are working on giving it a little refactoring to have the tests covering both ends. :)

ssdong replied:
Also, the current unit tests set meta for `requested` commit files which isn’t _strictly_ following and reflecting the actual behavior(we have empty requested file). I believe down the roadmap, we will have to standardize the requested file structure and investigate why it’s empty. There is a jira ticket filed for that and we have self-assigned and work on that next. :)

ssdong commented:
@satishkotha Thank you for your review! I’ll take a look when I get back. Currently on a day trip. 😄 
Basically, I wanna stop the abuse of `REQUESTED` here, at least for the insert overwrite writing operation, and separate it from `INFLIGHT`. With non-empty inflight commit files, we will suffer info loss. However, as you pointed out, this solution should also work against empty inflight files, I.e. clustering. 
I consider this a start to clean up and clarify various commit file logics as we have another issue of creating completely empty `REQUESTED` commit files.

ssdong commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/utils/MetadataConversionUtils.java:
```diff
@@ -72,10 +74,21 @@ public static HoodieArchivedMetaEntry createMetaWrapper(HoodieInstant hoodieInst
+          }
         } else {
-          HoodieRequestedReplaceMetadata requestedReplaceMetadata =
-              ClusteringUtils.getRequestedReplaceMetadata(metaClient, hoodieInstant).get();
```
Using `ClusteringUtils.getRequestedReplaceMetadata` is confusing to the reader even though it produces the same result. Part of the purpose of this method is to ignore the `inflight replacecommit` file and delegate it to `requested replacecommit` file to retrieve the clustering plan. However, what we merely need is a fetch&deserialization method for requested metadata. Hence, I have created a separate method for this single purpose.

ssdong commented on hudi-common/src/test/java/org/apache/hudi/common/testutils/HoodieTestTable.java:
```diff
@@ -239,14 +227,6 @@ public HoodieTestTable addRequestedCompaction(String instantTime, FileSlice... f
     return addRequestedCompaction(instantTime, plan);
   }
 
-  public HoodieTestTable addCompaction(String instantTime) throws IOException {
```
Cleaning up some unused methods.

ssdong commented on hudi-common/src/test/java/org/apache/hudi/common/testutils/FileCreateUtils.java:
```diff
@@ -157,11 +157,19 @@ public static void createReplaceCommit(String basePath, String instantTime, Hood
   public static void createRequestedReplaceCommit(String basePath, String instantTime, HoodieRequestedReplaceMetadata requestedReplaceMetadata) throws IOException {
-    createMetaFile(basePath, instantTime, HoodieTimeline.REQUESTED_REPLACE_COMMIT_EXTENSION, serializeRequestedReplaceMetadata(requestedReplaceMetadata).get());
+    if (null != requestedReplaceMetadata) {
+      createMetaFile(basePath, instantTime, HoodieTimeline.REQUESTED_REPLACE_COMMIT_EXTENSION, serializeRequestedReplaceMetadata(requestedReplaceMetadata).get());
+    } else {
+      createMetaFile(basePath, instantTime, HoodieTimeline.REQUESTED_REPLACE_COMMIT_EXTENSION, new byte[0]);
+    }
   }
 
-  public static void createInflightReplaceCommit(String basePath, String instantTime) throws IOException {
-    createMetaFile(basePath, instantTime, HoodieTimeline.INFLIGHT_REPLACE_COMMIT_EXTENSION);
+  public static void createInflightReplaceCommit(String basePath, String instantTime, HoodieCommitMetadata inflightReplaceMetadata) throws IOException {
+    if (inflightReplaceMetadata != null) {
+      createMetaFile(basePath, instantTime, HoodieTimeline.INFLIGHT_REPLACE_COMMIT_EXTENSION, inflightReplaceMetadata.toJsonString().getBytes(StandardCharsets.UTF_8));
+    } else {
+      createMetaFile(basePath, instantTime, HoodieTimeline.INFLIGHT_REPLACE_COMMIT_EXTENSION);
+    }
```
Both requested and inflight commit files can be empty and non-empty under existing implementations so we wanna test them both in the unit tests.

ssdong commented:
Hi @satishkotha, I've updated the PR to reflect the most up-to-date conversations regarding handling different scenarios of requested and inflight cases for insert_overwrite and clustering, and I've placed adequate comments to try not to confuse anyone who reads code, as it had confused me before. 
Also, I updated the PR description to attach the testing evidence from local. Please take a look. Unit tests have also been updated to cover both cases, and I did some trivial refactoring, e.g. remove unused methods.
Let me know if there is anything else we should be doing in this PR. Otherwise, it's in a good state to be reviewed again. Thanks!

zherenyu831 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/utils/MetadataConversionUtils.java:
```diff
@@ -76,15 +76,16 @@ public static HoodieArchivedMetaEntry createMetaWrapper(HoodieInstant hoodieInst
-                  ClusteringUtils.getRequestedReplaceMetadata(metaClient, hoodieInstant);
+          // we may have cases with empty HoodieRequestedReplaceMetadata e.g. insert_overwrite_table or insert_overwrite
+          // without clustering. However, we should revisit the requested commit file standardization
+          Option<HoodieRequestedReplaceMetadata> requestedReplaceMetadata = getRequestedReplaceMetadata(metaClient, hoodieInstant);
```
How about have a replaceUtils? I think maybe it is a good timing to decouple logic of replace commit from clustering

ssdong replied:
The `MetadataConversionUtils` should suffice to group such logic. They are all about _converting_ commit files to Avro schema based metadata. 🤔

satishkotha commented on hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/io/TestHoodieTimelineArchiveLog.java:
```diff
@@ -248,7 +252,7 @@ public void testArchiveTableWithReplacedFiles() throws Exception {
     HoodieTimeline timeline = metaClient.getActiveTimeline().getCommitsTimeline().filterCompletedInstants();
     HoodieTable table = HoodieSparkTable.create(cfg, context, metaClient);
-    assertEquals(4, timeline.countInstants(), "Loaded 4 commits and the count should match");
+    assertEquals(5, timeline.countInstants(), "Loaded 4 commits and the count should match");
```
nit: can you also update error message to say "Loaded 5 commits..."

satishkotha commented on hudi-common/src/test/java/org/apache/hudi/common/testutils/FileCreateUtils.java:
```diff
@@ -157,11 +157,19 @@ public static void createReplaceCommit(String basePath, String instantTime, Hood
+    if (null != requestedReplaceMetadata) {
+      createMetaFile(basePath, instantTime, HoodieTimeline.REQUESTED_REPLACE_COMMIT_EXTENSION, serializeRequestedReplaceMetadata(requestedReplaceMetadata).get());
+    } else {
+      createMetaFile(basePath, instantTime, HoodieTimeline.REQUESTED_REPLACE_COMMIT_EXTENSION, new byte[0]);
```
nit: i dont think you need  'new byte[0]'?

ssdong replied:
Nice catch. it isn't needed 😅

satishkotha reviewed (approved):
This looks much cleaner now. Thanks for making the change. LGTM.

ssdong commented:
@satishkotha Thank you for your reviewing! I've resolved the comments. 😄

susudong commented:
Weird. Didn’t think the trivial update would break the CI. Taking a look.

ssdong commented:
hmm, that's strange. I wonder why it would break the CI.. 
Taking out `new byte[0]` from `createMetaFile` should have 0 impact. It's supposed to create empty files anyway..
The error msg says:
```
Caused by: java.io.IOException: Not an Avro data file
	at org.apache.avro.file.DataFileReader.openReader(DataFileReader.java:50)
	at org.apache.hudi.common.table.timeline.TimelineMetadataUtils.deserializeAvroMetadata(TimelineMetadataUtils.java:178)
	at org.apache.hudi.common.util.CleanerUtils.getCleanerPlan(CleanerUtils.java:106)
	at org.apache.hudi.table.action.clean.BaseCleanActionExecutor.runPendingClean(BaseCleanActionExecutor.java:84)
```
Will take a look..

ssdong commented:
@satishkotha Actually, there is another PR #2809 failing upon `hudi-utilities` and `hudi-spark-client`at the same spot, `TestHoodieDeltaStreamer`, in the CI. I wonder if this is a global thing? 🤔

satishkotha commented:
> @satishkotha Actually, there is another PR #2809 failing upon `hudi-utilities` and `hudi-spark-client`at the same spot, `TestHoodieDeltaStreamer`, in the CI. I wonder if this is a global thing? 🤔

@ssdong possible failures are unrelated. Could you restart the build? I think you can push an empty commit to restart CI job. (You can also close and reopen PR)

ssdong commented:
> @ssdong possible failures are unrelated. Could you restart the build? I think you can push an empty commit to restart CI job. (You can also close and reopen PR)

Yep! Git allows empty commit. 😄 Triggered!

satishkotha commented:
@lw309637554 I reviewed this PR but just want to get another set of eyes. do you mind taking a pass on this?

ssdong commented:
@satishkotha @lw309637554 Just to share some updates, this PR fixed the following 2 issues during archival
1. `Positive number of partitions required`
2. `java.util.NoSuchElementException: No value present in Option`

However, the aforementioned 
```
// Initialize with new Hoodie timeline.
init(metaClient, getTimeline());
```
does cause
`java.io.FileNotFoundException: File file:/Users/susu.dong/Dev/clustering-insert-overwrite-test/.hoodie/20210415220131.replacecommit does not exist` which is a 3rd issue during archival
if we turn _off_ `"hoodie.clean.automatic"`, the cleaner option, which is `true` by default.

Turning off cleaner is making the internally maintained timeline to be out-of-sync with the physical commit file status. The archival removes the commit files while the `init` call still references those commit files that are being removed/archived when it propagates to the `readDataFromPath` method call and throws the exception ultimately.

Full stacktrace:
```
org.apache.hudi.exception.HoodieIOException: Could not read commit details from /Users/susu.dong/Dev/clustering-insert-overwrite-test/.hoodie/20210415220131.replacecommit
  at org.apache.hudi.common.table.timeline.HoodieActiveTimeline.readDataFromPath(HoodieActiveTimeline.java:561)
  at org.apache.hudi.common.table.timeline.HoodieActiveTimeline.getInstantDetails(HoodieActiveTimeline.java:225)
  at org.apache.hudi.common.table.view.AbstractTableFileSystemView.lambda$resetFileGroupsReplaced$8(AbstractTableFileSystemView.java:217)
  at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:271)
  at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654)
  at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
  at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
  at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
  at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
  at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
  at org.apache.hudi.common.table.view.AbstractTableFileSystemView.resetFileGroupsReplaced(AbstractTableFileSystemView.java:228)
  at org.apache.hudi.common.table.view.AbstractTableFileSystemView.init(AbstractTableFileSystemView.java:106)
  at org.apache.hudi.common.table.view.HoodieTableFileSystemView.init(HoodieTableFileSystemView.java:106)
  at org.apache.hudi.common.table.view.AbstractTableFileSystemView.reset(AbstractTableFileSystemView.java:248)
  at org.apache.hudi.common.table.view.HoodieTableFileSystemView.close(HoodieTableFileSystemView.java:353)
  at java.base/java.util.concurrent.ConcurrentHashMap$ValuesView.forEach(ConcurrentHashMap.java:4772)
  at org.apache.hudi.common.table.view.FileSystemViewManager.close(FileSystemViewManager.java:118)
  at org.apache.hudi.timeline.service.TimelineService.close(TimelineService.java:207)
  at org.apache.hudi.client.embedded.EmbeddedTimelineService.stop(EmbeddedTimelineService.java:119)
  at org.apache.hudi.client.AbstractHoodieClient.stopEmbeddedServerView(AbstractHoodieClient.java:94)
  at org.apache.hudi.client.AbstractHoodieClient.close(AbstractHoodieClient.java:86)
  at org.apache.hudi.client.AbstractHoodieWriteClient.close(AbstractHoodieWriteClient.java:1047)
  at org.apache.hudi.HoodieSparkSqlWriter$.commitAndPerformPostOperations(HoodieSparkSqlWriter.scala:505)
  at org.apache.hudi.HoodieSparkSqlWriter$.write(HoodieSparkSqlWriter.scala:225)
  at org.apache.hudi.DefaultSource.createRelation(DefaultSource.scala:161)
  at org.apache.spark.sql.execution.datasources.SaveIntoDataSourceCommand.run(SaveIntoDataSourceCommand.scala:46)
  at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult$lzycompute(commands.scala:70)
  at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult(commands.scala:68)
  at org.apache.spark.sql.execution.command.ExecutedCommandExec.doExecute(commands.scala:90)
  at org.apache.spark.sql.execution.SparkPlan.$anonfun$execute$1(SparkPlan.scala:175)
  at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:213)
  at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
  at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:210)
  at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:171)
  at org.apache.spark.sql.execution.QueryExecution.toRdd$lzycompute(QueryExecution.scala:122)
  at org.apache.spark.sql.execution.QueryExecution.toRdd(QueryExecution.scala:121)
  at org.apache.spark.sql.DataFrameWriter.$anonfun$runCommand$1(DataFrameWriter.scala:963)
  at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$5(SQLExecution.scala:100)
  at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:160)
  at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:87)
  at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:764)
  at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:64)
  at org.apache.spark.sql.DataFrameWriter.runCommand(DataFrameWriter.scala:963)
  at org.apache.spark.sql.DataFrameWriter.saveToV1Source(DataFrameWriter.scala:415)
  at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:399)
  at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:288)
```

We don't have to fix it in this PR, however, we do need to find a better solution without having to load the full active timeline that breaks some fundamental assumption. Is it possible for one of you to reach out to @bvaradar to get his attention on this? Thanks!

zherenyu831 commented:
There is one solution:
always enable cleaner, which will help you refresh timeline 😭

lw309637554 commented:
> @lw309637554 I reviewed this PR but just want to get another set of eyes. do you mind taking a pass on this?
okay

lw309637554 commented on hudi-common/src/test/java/org/apache/hudi/common/testutils/HoodieTestTable.java:
```diff
@@ -177,10 +161,14 @@ public HoodieTestTable addDeltaCommit(String instantTime) throws Exception {
+      String instantTime,
+      HoodieRequestedReplaceMetadata requestedReplaceMetadata,
+      HoodieReplaceCommitMetadata completeReplaceMetadata,
+      HoodieCommitMetadata inflightReplaceMetadata) throws Exception {
```
can we set HoodieCommitMetadata to Option< HoodieCommitMetadata >  to avoid null

ssdong replied:
hmm, if you track it down to `createInflightReplaceCommit` where `HoodieCommitMetadata` is being referenced, a `null` check is being presented there. However, I do agree an `Optional` reminds people of a null check though the
`o.isPresent()` check is _hardly_ any better than `o != null`. I could make the change and also update `createRequestedReplaceCommit` to adopt the same Optional `HoodieRequestedReplaceMetadata`.

lw309637554 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/ReplaceArchivalHelper.java:
```diff
@@ -68,6 +68,11 @@
                                                  HoodieInstant instant, List<String> replacedPartitions) {
+    // There is no file id to be replaced in the very first replace commit file for insert overwrite operation
+    if (replacedPartitions.isEmpty()) {
+      LOG.warn("Found empty partitionToReplaceFileIds");
```
can use  partitionToReplaceFileIds -> replacedPartitions better?

ssdong replied:
`partitionToReplaceFileIds` is the field name directly taken from the replace commit file. I guess it is subject to change. What about just explaining the warning as `Found no partition files to replace`?

lw309637554 replied:
yes, "Found no partition files to replace " will better

lw309637554 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/utils/MetadataConversionUtils.java:
```diff
@@ -105,14 +117,25 @@ public static HoodieArchivedMetaEntry createMetaWrapper(HoodieInstant hoodieInst
-    archivedMetaWrapper.setHoodieCommitMetadata(convertCommitMetadata(hoodieCommitMetadata));
-    archivedMetaWrapper.setActionType(ActionType.commit.name());
-    return archivedMetaWrapper;
+  public static Option<HoodieCommitMetadata> getInflightReplaceMetadata(HoodieTableMetaClient metaClient, HoodieInstant instant) throws IOException {
```
can we move this to ClusteringUtils

ssdong replied:
Hmm, I don't think this method is relevant to clustering.. 🤔They are solely responsible for retrieving corresponding commit files and parse them, which is more applicable to be residing in a "meta conversion" file, which comes within `MetadataConversionUtils.java` naturally. In fact, I've gotten rid of the dependency upon `ClusteringUtils` in `MetadataConversionUtils.java` after this refactoring. 😅

lw309637554 replied:
okay

lw309637554 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/utils/MetadataConversionUtils.java:
```diff
@@ -105,14 +117,25 @@ public static HoodieArchivedMetaEntry createMetaWrapper(HoodieInstant hoodieInst
+    return Option.of(HoodieCommitMetadata.fromBytes(inflightContent.get(), HoodieCommitMetadata.class));
+  }
+
+  public static Option<HoodieRequestedReplaceMetadata> getRequestedReplaceMetadata(HoodieTableMetaClient metaClient, HoodieInstant instant) throws IOException {
```
can we move this to ClusteringUtils

ssdong replied:
Same as above 😅

lw309637554 commented:
@zherenyu831 @ssdong thanks for your contribution, left some minor comments

ssdong commented:
@lw309637554 Thank you for your comments and I've replied. Please take a look and let me know. 😄

lw309637554 commented:
> @lw309637554 Thank you for your comments and I've replied. Please take a look and let me know. 😄

replied, can modify the log

ssdong commented:
@lw309637554 updated! Please take a look and let me know.

ssdong commented:
Hmmm, the build kept failing for random and irrelevant reasons. It might be better to have a “rerun” button somewhere so we could avoid the hassle of either re-opening or submit an empty commit just to trigger it...

lw309637554 commented:
LGTM

lw309637554 commented:
> LGTM, @satishkotha  can you check and merge it ?

vinothchandar commented:
I will take a pass on this and land!

satishkotha commented:
> I will take a pass on this and land!

@vinothchandar could you please review this since its been waiting for some time?

ssdong commented:
I've solved the conflicts..  let me know if you guys have second thoughts since this has been sitting for a bit while 😅

ssdong commented:
I'll fix the failing tests due to the merge of the master branch but we should seek to merge this PR sooner than later..

vinothchandar commented:
Thanks for this @ssdong and also the great explanation and detailed comments along the way. I am on it. This is a bit deep, been trying to set aside an hour or so. :)

ssdong commented:
@vinothchandar No problem! Thank you for your precious time and let me know if you have a second thought on this fix. :)

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/utils/MetadataConversionUtils.java:
```diff
@@ -107,14 +119,25 @@ public static HoodieArchivedMetaEntry createMetaWrapper(HoodieInstant hoodieInst
+  public static Option<HoodieCommitMetadata> getInflightReplaceMetadata(HoodieTableMetaClient metaClient, HoodieInstant instant) throws IOException {
+    Option<byte[]> inflightContent = metaClient.getActiveTimeline().getInstantDetails(instant);
+    if (!inflightContent.isPresent() || inflightContent.get().length == 0) {
+      // inflight files can be empty in some certain cases, e.g. when users opt in clustering
```
@satishkotha what is the original reason for this?

vinothchandar reviewed (approved):
LGTM overall.

fengjian428 commented:
@ssdong is there any quick fix can do in version 0.7.0?

ssdong commented:
@fengjian428 Can you try to build your customized Hudi with this fix here https://github.com/apache/hudi/pull/2784/files#diff-23f2c38220058fc24687f2cc138272e4b24e291e712e694dc122008f0672f2bd ?

fengjian428 commented:
> @fengjian428 Can you try to build your customized Hudi with this fix here https://github.com/apache/hudi/pull/2784/files#diff-23f2c38220058fc24687f2cc138272e4b24e291e712e694dc122008f0672f2bd ?

so I can just change ReplaceArchivalHelper part then it works ?
